source/server/listener_manager_impl.cc

```diff
@@ -1022,36 +982,50 @@ void ListenerManagerImpl::setNewOrDrainingSocketFactory(
   // the same address we are configured for and doesn't use SO_REUSEPORT. This is an edge case, but
   // may happen if a listener is removed and then added back with a same or different name and
   // intended to listen on the same address. This should work and not fail.
-  Network::ListenSocketFactorySharedPtr draining_listen_socket_factory;
+  const Network::ListenSocketFactory* draining_listen_socket_factory = nullptr;
   auto existing_draining_listener = std::find_if(
       draining_listeners_.cbegin(), draining_listeners_.cend(),
       [&listener](const DrainingListener& draining_listener) {
-        return draining_listener.listener_->listenSocketFactory().sharedSocket().has_value() &&
-               draining_listener.listener_->listenSocketFactory().sharedSocket()->get().isOpen() &&
+        return !draining_listener.listener_->listenSocketFactory().reusePort() &&
+               draining_listener.listener_->listenSocketFactory().getListenSocket(0)->isOpen() &&
                *listener.address() ==
                    *draining_listener.listener_->listenSocketFactory().localAddress();
       });
 
   if (existing_draining_listener != draining_listeners_.cend()) {
-    draining_listen_socket_factory = existing_draining_listener->listener_->getSocketFactory();
+    draining_listen_socket_factory = &existing_draining_listener->listener_->getSocketFactory();
   }
 
   Network::Socket::Type socket_type = Network::Utility::protobufAddressSocketType(proto_address);
   listener.setSocketFactory(
-      draining_listen_socket_factory
-          ? draining_listen_socket_factory
+      draining_listen_socket_factory != nullptr
+          ? draining_listen_socket_factory->clone()
```

Comment:
A high level question: do we support reuseport -> not-reuse, or not-reuse -> reuseport transition?

It can be done either by runtime flag or a real listener config update.

Reply:
NVM: reuse port impact the # of concurrency listen socket of the new listener config. The elder listener should not be impacted by the new value of `reuse_port`

Reply:
(Unresolving just to surface this conversation as I think it's important)

Actually, at the moment, I think any listener update that updates socket options (whether reuse port or something else) does not work correctly (the new options are dropped). This is because the current code (and this PR) use the existing sockets for the new listener. It's out of scope for this PR but should we warn about this case? I'm not sure blocking it makes sense.